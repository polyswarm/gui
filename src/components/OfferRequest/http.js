import multihashes from 'multihashes';
import validator from 'validator';
import etherutils from 'ethereumjs-util';
import web3Utils from 'web3-utils';
import Http from '../BountyCreate/http';

class HttpRequest {
  constructor(url) {
    this.url = url;
  }

  getUrlAccount(account) {
    return '?account=' + account;
  }

  uploadFiles(files) {
    const url = this.url;
    const helper = new Http(url);
    return helper.uploadFiles(files);
  }

  sendRequest(key, offer, artifactUri) {
    const url = this.url;
    return new Promise((resolve, reject) => {
      const hash = multihashes.fromB58String(artifactUri);
      try {
        multihashes.validate(hash);
        resolve(artifactUri);
      } catch (error) {
        reject(error);
      }
    }).then(() => new Promise((resolve, reject) => {
      if (validator.isUUID(offer.guid, 4)) {
        resolve();
      } else {
        reject('Invalid GUID');
      }
    })).then(() => new Promise((resolve, reject) => {
      if (web3Utils.isAddress(offer.ambassador)) {
        resolve();
      } else {
        reject();
      }
    })).then(() => new Promise(resolve => {
      const offerState = [];
      offerState.push(0); // is close
      offerState.push(0); // sequence
      offerState.push(0); // ambassador address
      offerState.push(0); // expert address
      offerState.push(0); //  msig address
      offerState.push(0); // new balance in nectar ambassador
      offerState.push(0); // balance in nectar expert
      offerState.push(0); // token address
      offerState.push(0); // A globally-unique identifer for the Listing.
      offerState.push(0); // The Offer Amount.
      offerState.push(0); // Cryptographic hash of the Artifact.
      offerState.push(artifactUri); // The URI of the Artifact.
      offerState.push(0); // Engagement Deadline
      offerState.push(0); // Assertion Deadline
      offerState.push(0); // has the expert made commitment
      offerState.push(0); // “malicious” or “benign”
      offerState.push(0); // Information derived during Assertion generation

      const depositState = this.marshallState(offerState);
      const buff_key = etherutils.toBuffer(key);
      const state = web3Utils.toHex(depositState);
      let msg = '0x' + etherutils.keccak(etherutils.toBuffer(state)).toString('hex');
      msg = '0x' + etherutils.hashPersonalMessage(etherutils.toBuffer(msg)).toString('hex');
      const sig = etherutils.ecsign(etherutils.toBuffer(msg), buff_key);
      let r = '0x' + sig.r.toString('hex');
      let s = '0x' + sig.s.toString('hex');
      let v = sig.v;

      const body = JSON.stringify({
        toSocketUri: offer.websocketUri,
        fromSocketUri: offer.websocketUri,
        state,
        v,
        r,
        s
      });
      resolve(body);
    })).then((body) => {
      const account = this.getUrlAccount(offer.ambassador);
      return fetch(url + '/offers/' + offer.guid + '/sendmsg' + account, {
        headers: {
          'Content-Type': 'application/json'
        },
        method: 'post',
        body: body
      });
    }).then(response => {
      if (response.ok) {
        return response;
      }
      return new Promise(resolve => {
        resolve(response.json());
      }).then(json => {
        throw Error(json.message);
      });
    }).then(response => response.json())
      .then(body => body.result);
  }

  getArtifactsList(uri) {
    const url = this.url;
    return new Promise((resolve, reject) => {
      const hash = multihashes.fromB58String(uri);
      try {
        multihashes.validate(hash);
        resolve(uri);
      } catch (error) {
        reject(error);
      }
    })
      .then(uri => fetch(url + '/artifacts/' + uri))
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(files => {
        return files.map(file => {
          file.name = file.name.trim();
          return file;
        });
      });
  }

  marshallState(inputs) {
    var m = this.getBytes(inputs[0]);

    for(var i=1; i<inputs.length;i++) {
      m += this.getBytes(inputs[i]).substr(2, this.getBytes(inputs[i]).length);
    }
    return m;
  }

  getBytes(input) {
    if(Buffer.isBuffer(input)) input = '0x' + input.toString('hex');
    if(66-input.length <= 0) return web3Utils.toHex(input);
    return this.padBytes32(web3Utils.toHex(input));
  }

  padBytes32(data){
    // TODO: check input is hex / move to TS
    let l = 66-data.length;

    let x = data.substr(2, data.length);

    for(var i=0; i<l; i++) {
      x = 0 + x;
    }
    return '0x' + x;
  }
}
export default HttpRequest;
