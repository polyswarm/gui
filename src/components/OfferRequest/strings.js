const strings = {
  request: 'Sending files.',
  error: 'Unable to send message to expert.',
};
export default strings;