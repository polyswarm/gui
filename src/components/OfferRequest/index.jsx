// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Uuid from 'uuid/v4';
// Bounty imports
import DropTarget from '../DropTarget';
import FileList from '../FileList';
import Button from '../Button';
import ModalPassword from '../ModalPassword';
// Component imports
import strings from './strings';
import Http from './http';

class OfferRequest extends Component {
  constructor(props) {
    super(props);
    this.state = {
      files: [],
    };
    this.onClearAll = this.onClearAll.bind(this);
    this.onClickHandler = this.onClickHandler.bind(this);
    this.onError = this.onError.bind(this);
    this.onFileRemoved = this.onFileRemoved.bind(this);
    this.onFilesSent = this.onFilesSent.bind(this);
    this.onMultipleFilesSelected = this.onMultipleFilesSelected.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
    this.addMessage = this.addMessage.bind(this);
    this.addSendMessageRequest = this.addSendMessageRequest.bind(this);
    this.removeSendMessageRequest = this.removeSendMessageRequest.bind(this);
    this.sendMessage = this.sendMessage.bind(this);
  }

  componentDidMount() {
    const { props: { url } } = this;
    this.http = new Http(url);
  }

  render() {
    const { state: { files } } = this;
    const { props: { url, offer, addRequest, removeRequest } } = this;
    const walletList = [offer.author];
    return (
      <div className='Bounty-Create'>
        <ModalPassword
          ref={modal => (this.modal = modal)}
          url={url}
          walletList={walletList}
          onWalletChange={this.onWalletChangeHandler}
          addRequest={addRequest}
          removeRequest={removeRequest}/>
        <div className='OfferRequest-Content'>
          <div className='Bounty-Files'>
            <DropTarget onFilesSelected={this.onMultipleFilesSelected} />
            <FileList
              files={files}
              clear={this.onClearAll}
              removeFile={this.onFileRemoved}/>
            <div className='Bounty-Create-Upload'>
              <Button
                disabled={ !files || files.length == 0 }
                onClick={this.onClickHandler}>
                {`Send ${files.length} files.`}
              </Button>
            </div>
          </div>
        </div>
      </div>
    );
  }

  onClearAll() {
    this.setState({ files: []});
  }
  
  onClickHandler() {
    this.modal.open();
  }
  
  onError(message) {
    const { props: { onError } } = this;
    if (onError) {
      onError(message);
    }
  }
  
  onFileRemoved(index) {
    const files = this.state.files.slice();
    if (index >= 0 && index < files.length) {
      files.splice(index, 1);
      this.setState({ files: files });
    }
  }
  
  onFilesSent() {
    const {props: {onFilesSent}} = this;
    if (onFilesSent) {
      onFilesSent();
    }
  }

  onMultipleFilesSelected(files) {
    const f = this.state.files.slice();
    const combined = f.concat(files);
    this.setState({ files: combined });
  }
  
  onWalletChangeHandler(didUnlock) {
    if (didUnlock) {
      this.createBounty();
    }
  }

  addMessage(message) {
    const {props: {addMessage}} = this;
    if (addMessage) {
      addMessage(message);
    }
  }

  addSendMessageRequest(id) {
    const { addRequest } = this.props;
    if (addRequest) {
      addRequest(strings.request, id);
    }
  }

  removeSendMessageRequest(id) {
    const { removeRequest } = this.props;
    if (removeRequest) {
      removeRequest(strings.request, id);
    }
  }

  sendMessage() {
    const files = this.state.files.slice();

    const http = this.http;
    if (files && files.length > 0) {
      const uuid = Uuid();
      this.addSendMessageRequest(uuid);
      this.setState({ files: [] });
      return new Promise((resolve) => {
        this.onFilesSent();
        resolve();
      })
        .then(() => http.uploadFiles(files))
        .then(artifact => http.sendMessage(artifact))
        .then(result => this.addMessage(result))
        .catch(error => {
          let errorMessage;
          if (!error || !error.message || error.message.length === 0) {
            errorMessage = strings.error;
          } else {
            errorMessage = error.message;
          }

          //Update app
          this.onError(errorMessage);
        })
        .then(() => {
          this.removeSendMessageRequest(uuid);
        });
    } else {
      return null;
    }
  }
}

OfferRequest.propTypes = {
  addMessage: PropTypes.func,
  addRequest: PropTypes.func,
  onError: PropTypes.func,
  removeRequest: PropTypes.func,
  url: PropTypes.string
};
export default OfferRequest;
