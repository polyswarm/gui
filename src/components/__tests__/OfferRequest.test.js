import React from 'react';
import {render} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import OfferRequest from '../OfferRequest';

it('renders without crashing', () => {
  const wrapper = render(<OfferRequest />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});