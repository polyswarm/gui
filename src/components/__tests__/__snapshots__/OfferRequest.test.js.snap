// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="OfferRequest"
>
  <header
    class="Header"
  >
    <div
      class="Header-Row"
    >
      <div
        class="Header-Title"
      >
        <img
          alt="PolySwarm logo"
          class="Header-Logo"
          src="../public/img/back-arrow.svg"
        >
          
        </img>
        <h3>
          Send files to expert
        </h3>
        <span
          class="RequestSpinner"
        >
          
        </span>
      </div>
      <div
        class="Header-Actions"
      >
        
      </div>
    </div>
    <div
      class="Header-Address"
    >
      <p
        class="Header-Divided"
        style="font-size:7.0588235294117645vw"
      >
        author
      </p>
      <p
        class="Header-Divided"
        style="font-size:7.0588235294117645vw"
      >
        1 NCT 
      </p>
      <p
        style="font-size:7.0588235294117645vw"
      >
        1 ETH
      </p>
    </div>
  </header>
  <div
    class="OfferRequest-Content"
  >
    <div
      class="OfferRequest-Files"
    >
      <h2>
        Select files to send to expert.
      </h2>
      <div
        class="OfferRequest-Button"
      >
        <button
          class="Button"
          disabled=""
        >
          Send 0 files.
        </button>
      </div>
      <div
        class="Drop-Target"
      >
        <div>
          <p>
            Drag and drop your files here, or 
          </p>
          <form>
            <input
              class="hidden"
              id="file"
              multiple=""
              type="file"
            >
              
            </input>
            <label
              class="LabelButton"
              for="file"
            >
              click here.
            </label>
          </form>
        </div>
      </div>
      <div
        class="File-List"
      >
        <header
          class="File-List-Header"
        >
          <p
            class="File-List-Title"
          >
            0 Files
          </p>
          <button
            class="Button flat-cancel"
            disabled=""
          >
            Clear
          </button>
        </header>
        <ul
          class="List"
        >
          <div>
            
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
